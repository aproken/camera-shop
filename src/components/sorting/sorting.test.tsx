import { render, RenderResult, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Sorting from './sorting';

describe('Sorting компонент', () => {
  it('Отрисовка компонента', () => {
    const mockOnChange = jest.fn((sortByType: string, sortByOrder: string) => null);

    const { container }: RenderResult = render(<Sorting onChange={ mockOnChange }/>);

    expect(container).toMatchSnapshot();
  });

  it('Реагирует на действия пользователя', () => {
    const mockOnChange = jest.fn((sortByType: string, sortByOrder: string) => null);

    render(<Sorting onChange={ mockOnChange }/>);

    const inputSortByTypePrice = screen.getByRole('radio', {name: 'по цене'});
    const inputSortByTypePopularity = screen.getByRole('radio', {name: 'по популярности'});
    const inputSortByOrderAsc = screen.getByRole('radio', {name: 'По возрастанию'});
    const inputSortByOrderDesc = screen.getByRole('radio', {name: 'По убыванию'});

    fireEvent.click(inputSortByTypePrice);
    expect(inputSortByTypePrice).toBeChecked();
    expect(inputSortByTypePopularity).not.toBeChecked();

    fireEvent.click(inputSortByTypePopularity);
    expect(inputSortByTypePrice).not.toBeChecked();
    expect(inputSortByTypePopularity).toBeChecked();

    fireEvent.click(inputSortByOrderAsc);
    expect(inputSortByOrderAsc).toBeChecked();
    expect(inputSortByOrderDesc).not.toBeChecked();

    fireEvent.click(inputSortByOrderDesc);
    expect(inputSortByOrderAsc).not.toBeChecked();
    expect(inputSortByOrderDesc).toBeChecked();
  });

  it('Корректно передает значения в коллбэк', () => {
    const mockOnChange = jest.fn((sortByType: string, sortByOrder: string) => null);

    render(<Sorting onChange={ mockOnChange }/>);

    const inputSortByTypePrice = screen.getByRole('radio', {name: 'по цене'});
    const inputSortByTypePopularity = screen.getByRole('radio', {name: 'по популярности'});
    const inputSortByOrderAsc = screen.getByRole('radio', {name: 'По возрастанию'});
    const inputSortByOrderDesc = screen.getByRole('radio', {name: 'По убыванию'});

    fireEvent.click(inputSortByTypePrice);
    expect(mockOnChange).toBeCalledWith('по цене', '');

    mockOnChange.mockClear();

    fireEvent.click(inputSortByTypePopularity);
    expect(mockOnChange).toBeCalledWith('по полулярности', '');

    mockOnChange.mockClear();

    fireEvent.click(inputSortByTypePrice);
    fireEvent.click(inputSortByOrderAsc);
    expect(mockOnChange).toBeCalledWith('по цене', 'abs');

    mockOnChange.mockClear();

    fireEvent.click(inputSortByTypePopularity);
    fireEvent.click(inputSortByOrderDesc);
    expect(mockOnChange).toBeCalledWith('по полулярности', 'desc');
  });
});
