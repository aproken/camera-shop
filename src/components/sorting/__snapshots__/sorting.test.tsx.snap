// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sorting компонент Отрисовка компонента 1`] = `
<div>
  <form
    action="#"
  >
    <div
      class="catalog-sort__inner"
    >
      <p
        class="title title--h5"
      >
        Сортировать:
      </p>
      <div
        class="catalog-sort__type"
      >
        <div
          class="catalog-sort__btn-text"
        >
          <input
            id="по цене"
            name="sort"
            type="radio"
          />
          <label
            for="по цене"
          >
            по цене
          </label>
        </div>
        <div
          class="catalog-sort__btn-text"
        >
          <input
            id="по полулярности"
            name="sort"
            type="radio"
          />
          <label
            for="по полулярности"
          >
            по популярности
          </label>
        </div>
      </div>
      <div
        class="catalog-sort__order"
      >
        <div
          class="catalog-sort__btn catalog-sort__btn--up"
        >
          <input
            aria-label="По возрастанию"
            id="abs"
            name="sort-icon"
            type="radio"
          />
          <label
            for="abs"
          >
            <svg
              aria-hidden="true"
              height="14"
              width="16"
            >
              <use
                xlink:href="#icon-sort"
              />
            </svg>
          </label>
        </div>
        <div
          class="catalog-sort__btn catalog-sort__btn--down"
        >
          <input
            aria-label="По убыванию"
            id="desc"
            name="sort-icon"
            type="radio"
          />
          <label
            for="desc"
          >
            <svg
              aria-hidden="true"
              height="14"
              width="16"
            >
              <use
                xlink:href="#icon-sort"
              />
            </svg>
          </label>
        </div>
      </div>
    </div>
  </form>
</div>
`;
